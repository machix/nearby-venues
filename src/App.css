.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 120px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}


.venue-marker {
  padding: 5px;
  position: relative;
  color: white;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
  width: 70px;
  left: -50%;
  transform: translateY(-100%);
  display: flex;
  align-items: center;
  justify-content: space-between;
  z-index: 1;
}

.venue-rating {
  border: 0.5px solid white;
  border-radius: 2px;
  padding: 2px 5px;
}

.venue-tier {
}

.bottom-arrow {
  position: absolute;
  left: calc(50% - 5px);
  top: 100%;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #000;
  clear: both;
}
.venue-marker:hover {
  z-index: 99;
  background-color: orange!important;
}
.venue-marker:hover .bottom-arrow {
  border-top-color: orange!important;
}
.venue-marker:hover .marker-details{
  display: block;
  padding: 10px;
  opacity: 1;
  visibility: visible;
}
.marker-details {
  visibility: hidden;
  position: absolute;
  background-color: white;
  color: black;
  opacity: 0;
  bottom: 100%;
  left: calc(50% - 125px);
  box-sizing: border-box;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
  width: 250px;
  overflow: hidden;
  padding: 0;
  transition: opacity 0.5s, padding 0.2s;
}
.venue-name {
  font-weight: bold;
    margin: 0;
    padding-bottom: 5px;
    border-bottom: 1px solid #e2e2e2;
    margin-bottom: 5px;
    display: block;
    color: black;
}
.radius-select {
  position: absolute;
  right: 10px;
  top: 10px;  
}
.marker-detail-content {
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.marker-detail-content .left-content {
  display: flex;
  flex-direction: column;
  align-items: space-between;
}
.venue-tag {
    padding: 5px;
    margin: 0;
    background-color: orange;
    border-radius: 30px;
    color: white;
}

.address-text {
  margin: 5px;
}

.loader {
      position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: #00000042;
}

/* map */
.map-element {
  height: calc(100vh - 175px);
}

.full-height {
  height: 100%;
}

/* button */


.btn {
  position: relative;
  display: block;
  padding: 0;
  overflow: hidden;
  border-width: 0;
  outline: none;
  border-radius: 2px;
  box-shadow: 0 1px 4px rgba(0, 0, 0, .6);
  background-color: #00A7F5;
  color: #ecf0f1;  
  transition: background-color .3s;
  margin: 5px auto;
  width: 100%;
}

.btn:hover, .btn:focus {
  background-color: #0186D4;
}

.btn > * {
  position: relative;
}

.btn span {
  display: block;
  padding: 5px 10px;
}

.btn:before {
  content: "";
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  background-color: rgba(236, 240, 241, .3);
  transform: translate(-50%, -50%);
}

.btn:active:before {
  width: 120%;
  padding-top: 120%;
  transition: width .2s ease-out, padding-top .2s ease-out;
}

/* Styles, not important */
*, *:before, *:after {
  box-sizing: border-box;
}
a, a:hover {
  text-decoration: none;
}
