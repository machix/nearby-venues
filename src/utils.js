function getQueryString(params) {
  var esc = encodeURIComponent;
  return Object.keys(params)
    .map(k => esc(k) + '=' + esc(params[k]))
    .join('&');
}

const defaultVenueOptions = {
  client_id: '1CPAVHCZ44EOTXDVRLPYJPWXBSAVTNSWGVLG0QRST32H3DNR',
  client_secret: '51MYE2M4S0NJIBG5YOVQLTW1RS00XEV0UZD32BPYYNDYWTJI',
  openNow: 1,
  v: '20170801',
}

const fetchVenues = ({ lat, lng, radius }) => {
  const qs = {
    ...defaultVenueOptions,
    ll: `${lat},${lng}`,
    radius
  }
  return fetch('https://api.foursquare.com/v2/venues/explore?' + getQueryString(qs))
    .then(res => res.json())
    .then(({ response }) => {
      const venues = response.groups[0].items.map(({ venue }) => ({
        formattedAddress: venue.location.formattedAddress,
        phone: venue.contact.phone,
        name: venue.name,
        tier: venue.price && venue.price.tier,
        rating: venue.rating,
        url: venue.url,
        ratingColor: venue.ratingColor,
        lat: venue.location.lat,
        lng: venue.location.lng,
        tag: venue.categories[0].name
      }));
      return venues;
    })
};

export default {fetchVenues};