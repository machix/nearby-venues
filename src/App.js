import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import { GoogleMap, Marker, withGoogleMap, withScriptjs, OverlayView} from "react-google-maps"

function getQueryString(params) {
  var esc = encodeURIComponent;
  return Object.keys(params)
    .map(k => esc(k) + '=' + esc(params[k]))
    .join('&');
}

const VenueMarker = ({ rating, tier, ratingColor }) => (
  <div style={{ backgroundColor: '#' + ratingColor}} class='venue-marker'>
    {rating}
    {' | ' + new Array(tier).fill('$').join(' ')}
    <span class='bottom-arrow' style={{ borderTopColor: '#' + ratingColor }}></span>
  </div>
);

const Map = withScriptjs(withGoogleMap(({ defaultCenter, venues}) => <GoogleMap
  defaultZoom={16}
  center={defaultCenter}
>
  <Marker position={defaultCenter} />
  {venues.map((venue, i) => <OverlayView mapPaneName={OverlayView.OVERLAY_MOUSE_TARGET} key={i} position={venue} ><VenueMarker {...venue}/></OverlayView>)}
</GoogleMap>))

class App extends Component {
  state = {
    venues: [],
    radius: '',
    currentLocation: { lat: 40.7243, lng: -74.0018}
  }

  fetchVenues = ({lat, lng}) => {
    const qs = {
      client_id: '1CPAVHCZ44EOTXDVRLPYJPWXBSAVTNSWGVLG0QRST32H3DNR',
      client_secret: '51MYE2M4S0NJIBG5YOVQLTW1RS00XEV0UZD32BPYYNDYWTJI',
      openNow: 1,
      ll: `${lat},${lng}`,
      v: '20170801'
    }
    fetch('https://api.foursquare.com/v2/venues/explore?' + getQueryString(qs))
      .then(res => res.json())
      .then(({response}) => {
        const venues = response.groups[0].items.map(({venue}) => ({
          formattedAddress: venue.location.formattedAddress,
          contact: venue.contact,
          name: venue.name,
          tier: venue.price && venue.price.tier,
          rating: venue.rating,
          ratingColor: venue.ratingColor,
          lat: venue.location.lat,
          lng: venue.location.lng,
          tag: venue.categories[0].name
        }));
        this.setState({venues});
        console.log('hehehhe', venues);
      })
  }
  componentDidMount() {
    navigator.geolocation.getCurrentPosition((position) => {
      const currentLocation = { lat: position.coords.latitude, lng: position.coords.longitude };
      this.setState({ currentLocation});
      this.fetchVenues(currentLocation);
      console.log('position', position);
    }, console.log );
    this.fetchVenues({ lat: 40.7243, lng: -74.0018});    
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Welcome to React</h1>
        </header>
        <p className="App-intro">
          To get started, edit <code>src/App.js</code> and save to reload.
        </p>
        {this.state.currentLocation.lat ? 
          <Map defaultCenter={this.state.currentLocation}
            googleMapURL="https://maps.googleapis.com/maps/api/js?v=3.exp&libraries=geometry,drawing,places"
            loadingElement={<div style={{ height: `100%` }} />}
            containerElement={<div style={{ height: `400px` }} />}
            venues={this.state.venues}
            mapElement={<div style={{ height: `100%` }} />}></Map> : <p>fetching location</p>}
        
      </div>
    );
  }
}

export default App;
