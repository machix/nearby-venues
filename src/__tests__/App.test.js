import React from 'react';
import App from '../App';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import { wrap } from 'module';
import utils from '../utils';


jest.mock('../utils', () => ({
  fetchVenues: jest.fn(() => Promise.resolve(['1', '2']))
}))
describe('App.js: contains logic of fetching data', () => {
  let wrapper, instance;
  beforeEach(() => {
    wrapper = shallow(<App/>);
    instance = wrapper.instance();
  })
  it('renders correctly', () => {
    const tree = renderer.create(<App />);
    expect(tree).toMatchSnapshot();
  });
  it('findNearbyVenues: success', async () => {
    wrapper.setState({ radius: '250'});
    await instance.findNearbyVenues({
      lat: 13.7563,
      lng: 100.5018});
    expect(utils.fetchVenues).toBeCalledWith({
      lat: 13.7563,
      lng: 100.5018,
      radius: '250'
    })
    expect(wrapper.state().venues).toEqual(['1', '2'])
  })
  it('findNearbyVenues: failure', async () => {
    utils.fetchVenues.mockImplementation(() => Promise.reject('api failed'))
      await instance.findNearbyVenues({
        lat: 13.7563,
        lng: 100.5018
      });
    expect(alert).toBeCalledWith('Error fetching venues');
  })

  it('componentDidMount: should fetch user\'s location', () => {
    instance.findNearbyVenues = jest.fn();
    instance.componentDidMount();
    expect(navigator.geolocation.getCurrentPosition).toBeCalled()
    expect(instance.findNearbyVenues).toBeCalledWith({
      lat: 13.7563,
      lng: 100.5018
    })
  })
  it('changeRadius: should set radius and zoom level', () => {
    instance.findNearbyVenues = jest.fn();
    instance.changeRadius({target: {value: '500'}});
    expect(instance.state.radius).toBe('500');
    expect(instance.state.zoom).toBe(16);
    expect(instance.findNearbyVenues).toBeCalledWith({
      lat: 13.7563,
      lng: 100.5018
    })
  })
})

