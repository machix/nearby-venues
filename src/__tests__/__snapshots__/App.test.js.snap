// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.js: contains logic of fetching data renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src={
        Object {
          "process": [Function],
        }
      }
    />
  </header>
  <p
    className="App-intro"
  >
    <code>
      Here are some things to do near you. Tap on any marker to know more.
    </code>
    <br />
    <code>
      Radius 
      <select
        onChange={[Function]}
        value="250"
      >
        <option
          value="250"
        >
          250
           meters
        </option>
        <option
          value="500"
        >
          500
           meters
        </option>
        <option
          value="1000"
        >
          1000
           meters
        </option>
        <option
          value="2000"
        >
          2000
           meters
        </option>
        <option
          value="3000"
        >
          3000
           meters
        </option>
        <option
          value="5000"
        >
          5000
           meters
        </option>
        <option
          value="10000"
        >
          10000
           meters
        </option>
      </select>
    </code>
  </p>
  <div
    className="full-height"
  />
  <div
    className="loader"
  >
    <div>
      <svg
        fill="white"
        height="100"
        viewBox="0 0 135 140"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="120"
          rx="6"
          width="15"
          y="10"
        >
          <animate
            attributeName="height"
            begin="0.5s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="120;110;100;90;80;70;60;50;40;140;120"
          />
          <animate
            attributeName="y"
            begin="0.5s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="10;15;20;25;30;35;40;45;50;0;10"
          />
        </rect>
        <rect
          height="120"
          rx="6"
          width="15"
          x="30"
          y="10"
        >
          <animate
            attributeName="height"
            begin="0.25s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="120;110;100;90;80;70;60;50;40;140;120"
          />
          <animate
            attributeName="y"
            begin="0.25s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="10;15;20;25;30;35;40;45;50;0;10"
          />
        </rect>
        <rect
          height="140"
          rx="6"
          width="15"
          x="60"
        >
          <animate
            attributeName="height"
            begin="0s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="120;110;100;90;80;70;60;50;40;140;120"
          />
          <animate
            attributeName="y"
            begin="0s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="10;15;20;25;30;35;40;45;50;0;10"
          />
        </rect>
        <rect
          height="120"
          rx="6"
          width="15"
          x="90"
          y="10"
        >
          <animate
            attributeName="height"
            begin="0.25s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="120;110;100;90;80;70;60;50;40;140;120"
          />
          <animate
            attributeName="y"
            begin="0.25s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="10;15;20;25;30;35;40;45;50;0;10"
          />
        </rect>
        <rect
          height="120"
          rx="6"
          width="15"
          x="120"
          y="10"
        >
          <animate
            attributeName="height"
            begin="0.5s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="120;110;100;90;80;70;60;50;40;140;120"
          />
          <animate
            attributeName="y"
            begin="0.5s"
            calcMode="linear"
            dur="1s"
            repeatCount="indefinite"
            values="10;15;20;25;30;35;40;45;50;0;10"
          />
        </rect>
      </svg>
    </div>
  </div>
</div>
`;
